tileserver/__init__.py

```diff
@@ -221,14 +224,36 @@ def handle_request(self, request):
         format = request_data.format
 
         if self.store and coord.zoom <= 20:
-            # we have a dynamic layer request
-            # in this case, we should try to fetch the data from the
-            # cache, and if present, prune the layers that aren't
-            # necessary from there.
+            # we either have a dynamic layer request, or it's a
+            # request for a new tile that is not currently in the
+            # tiles of interest, or it's for a request that's in the
+            # tiles of interest that hasn't been generated, possibly
+            # because a new prefix is used and all tiles haven't been
+            # generated yet before making the switch
+
+            # in any case, it makes sense to try and fetch the json
+            # format from the store first
             tile_data = self.store.read_tile(coord, json_format)
             if tile_data is not None:
+                # the json format exists in the store
+                # we'll use it to generate the response
                 tile_data = reformat_selected_layers(tile_data, layer_data,
-                        coord, format)
+                                                     coord, format)
+                if layer_spec == 'all':
+                    # for the all layer, since the json format
+                    # existed, we should also save the requested
+                    # format too to allow the caches to serve it
+                    # directly in subsequent requests
```

Comment:
It seems like a bit of an edge case though, because if the original request was for json, shouldn't it have been served by the cache directly? But I think it still makes sense to add it in case of any race conditions and just to make the code more explicit. Updated with af86954.

I'm assuming that we still want to enqueue always though to be safe.